psi4/src/psi4/libmints/overlap.cc

```diff
@@ -28,372 +28,41 @@
+#include <libint2/engine.h>
 
-;
 using namespace psi;
 
 OverlapInt::OverlapInt(std::vector<SphericalTransform> &st, std::shared_ptr<BasisSet> bs1,
                        std::shared_ptr<BasisSet> bs2, int deriv)
-    : OneBodyAOInt(st, bs1, bs2, deriv), overlap_recur_(bs1->max_am() + deriv, bs2->max_am() + deriv) {
-    int maxam1 = bs1_->max_am();
-    int maxam2 = bs2_->max_am();
-
+    : OneBodyAOInt(st, bs1, bs2, deriv) {
     if (deriv > 2) {
         throw std::runtime_error("OverlapInt: does not support 3rd order derivatives and higher.");
     }
 
-    int maxnao1 = INT_NCART(maxam1);
-    int maxnao2 = INT_NCART(maxam2);
+    int max_am = std::max(basis1()->max_am(), basis2()->max_am());
+    int max_nprim = std::max(basis1()->max_nprimitive(), basis2()->max_nprimitive());
+
+    engine0_ = std::make_unique<libint2::Engine>(libint2::Operator::overlap, max_nprim, max_am, 0);
 
     if (deriv == 1) {
         // We set chunk count for normalize_am and pure_transform
         set_chunks(6);
 
-        maxnao1 *= 3;
-        maxnao2 *= 3;
+        engine1_ = std::make_unique<libint2::Engine>(libint2::Operator::overlap, max_nprim, max_am, 1);
     } else if (deriv == 2) {
-        set_chunks(6);
-        maxnao1 *= 6;
-        maxnao2 *= 1;
-    }
-
-    buffer_ = new double[maxnao1 * maxnao2];
-}
-
-OverlapInt::~OverlapInt() { delete[] buffer_; }
-
-// The engine only supports segmented basis sets
-void OverlapInt::compute_pair(const GaussianShell &s1, const GaussianShell &s2) {
-    int ao12;
-    int am1 = s1.am();
-    int am2 = s2.am();
-    int nprim1 = s1.nprimitive();
-    int nprim2 = s2.nprimitive();
-    double A[3], B[3];
-    A[0] = s1.center()[0];
-    A[1] = s1.center()[1];
-    A[2] = s1.center()[2];
-    B[0] = s2.center()[0];
-    B[1] = s2.center()[1];
-    B[2] = s2.center()[2];
-
-    // compute intermediates
-    double AB2 = 0.0;
-    AB2 += (A[0] - B[0]) * (A[0] - B[0]);
-    AB2 += (A[1] - B[1]) * (A[1] - B[1]);
-    AB2 += (A[2] - B[2]) * (A[2] - B[2]);
-
-    memset(buffer_, 0, s1.ncartesian() * s2.ncartesian() * sizeof(double));
-
-    double **x = overlap_recur_.x();
-    double **y = overlap_recur_.y();
-    double **z = overlap_recur_.z();
-
-    for (int p1 = 0; p1 < nprim1; ++p1) {
-        double a1 = s1.exp(p1);
-        double c1 = s1.coef(p1);
-        for (int p2 = 0; p2 < nprim2; ++p2) {
-            double a2 = s2.exp(p2);
-            double c2 = s2.coef(p2);
-            double gamma = a1 + a2;
-            double oog = 1.0 / gamma;
-
-            double PA[3], PB[3];
-            double P[3];
-
-            P[0] = (a1 * A[0] + a2 * B[0]) * oog;
-            P[1] = (a1 * A[1] + a2 * B[1]) * oog;
-            P[2] = (a1 * A[2] + a2 * B[2]) * oog;
-            PA[0] = P[0] - A[0];
-            PA[1] = P[1] - A[1];
-            PA[2] = P[2] - A[2];
-            PB[0] = P[0] - B[0];
-            PB[1] = P[1] - B[1];
-            PB[2] = P[2] - B[2];
+        set_chunks(21);
```

Comment:
21 chunks? How does somebody guess that magic number?

Reply:
It's the number of nonzero second derivatives - there's a good explanation on the libint2 wiki, as well as the JCC article that first introduces the CCA standard.